File: client-apps/cli/pkg/ignore/matcher.go
package ignore

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/go-git/go-git/v5/plumbing/format/gitignore"
)

// Options configures how the Matcher is constructed.
type Options struct {
	// RootDir is the artifact root directory (required).
	// This is where .gitignore and .stigmerignore files are looked for.
	RootDir string

	// RespectGitignore enables loading patterns from .gitignore.
	// Defaults to true if not explicitly set via NewWithDefaults.
	RespectGitignore bool

	// IncludeDefaults enables built-in security patterns.
	// Defaults to true if not explicitly set via NewWithDefaults.
	IncludeDefaults bool

	// ExtraIgnore contains additional patterns to ignore (e.g., from CLI --ignore).
	ExtraIgnore []string

	// ExtraInclude contains patterns to force-include (e.g., from CLI --include).
	// These are converted to negation patterns and have highest priority.
	ExtraInclude []string
}

// Matcher determines whether paths should be ignored during artifact creation.
// It composes multiple pattern sources with proper precedence ordering.
//
// Matcher is safe for concurrent use after creation.
type Matcher struct {
	// patterns contains all patterns in priority order (lowest to highest).
	// The last matching pattern wins.
	patterns []patternEntry

	// rootDir is the artifact root directory.
	rootDir string
}

// New creates a Matcher with the given options.
//
// Returns an error if RootDir is empty or doesn't exist.
// Pattern parse errors are silently ignored (lenient like Git).
//
// The default behavior is:
//   - RespectGitignore: false (must be explicitly enabled)
//   - IncludeDefaults: false (must be explicitly enabled)
//
// Use NewWithDefaults for standard behavior with security defaults enabled.
func New(opts Options) (*Matcher, error) {
	if opts.RootDir == "" {
		return nil, fmt.Errorf("RootDir is required")
	}

	// Normalize and validate RootDir
	rootDir, err := filepath.Abs(opts.RootDir)
	if err != nil {
		return nil, fmt.Errorf("invalid RootDir %q: %w", opts.RootDir, err)
	}

	info, err := os.Stat(rootDir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("RootDir does not exist: %s", rootDir)
		}
		return nil, fmt.Errorf("cannot access RootDir %s: %w", rootDir, err)
	}
	if !info.IsDir() {
		return nil, fmt.Errorf("RootDir is not a directory: %s", rootDir)
	}

	// Build patterns in precedence order (lowest to highest priority)
	var patterns []patternEntry

	// 1. Built-in defaults (lowest priority)
	if opts.IncludeDefaults {
		patterns = append(patterns, loadDefaultPatterns()...)
	}

	// 2. .gitignore
	if opts.RespectGitignore {
		patterns = append(patterns, loadGitignore(rootDir)...)
	}

	// 3. .stigmerignore
	patterns = append(patterns, loadStigmerignore(rootDir)...)

	// 4. CLI patterns (highest priority)
	if len(opts.ExtraIgnore) > 0 || len(opts.ExtraInclude) > 0 {
		patterns = append(patterns, loadCLIPatterns(opts.ExtraIgnore, opts.ExtraInclude)...)
	}

	return &Matcher{
		patterns: patterns,
		rootDir:  rootDir,
	}, nil
}

// NewWithDefaults creates a Matcher with security defaults enabled.
// This is the recommended constructor for typical use cases.
//
// Equivalent to:
//
//	New(Options{
//	    RootDir:          rootDir,
//	    RespectGitignore: true,
//	    IncludeDefaults:  true,
//	})
func NewWithDefaults(rootDir string) (*Matcher, error) {
	return New(Options{
		RootDir:          rootDir,
		RespectGitignore: true,
		IncludeDefaults:  true,
	})
}

// Match returns true if the path should be ignored (excluded from the artifact).
//
// The path must be relative to the artifact root and should use forward slashes.
// isDir should be true if the path represents a directory.
//
// This is the hot path for filtering - optimized for minimal allocations.
func (m *Matcher) Match(path string, isDir bool) bool {
	result := m.MatchWithReason(path, isDir)
	return result.Ignored
}

// MatchWithReason returns detailed information about the match decision.
// This is useful for debugging, dry-run mode, and understanding ignore behavior.
//
// The path must be relative to the artifact root and should use forward slashes.
// isDir should be true if the path represents a directory.
func (m *Matcher) MatchWithReason(path string, isDir bool) MatchResult {
	result := MatchResult{
		Path:  path,
		IsDir: isDir,
	}

	// Normalize path
	path = normalizePath(path)
	if path == "" {
		// Root directory is never ignored
		return result
	}

	// Convert to path components for go-git
	components := pathToComponents(path)
	if len(components) == 0 {
		return result
	}

	// Evaluate patterns in order - last match wins
	var lastMatch *patternEntry
	var lastMatchResult gitignore.MatchResult

	for i := range m.patterns {
		entry := &m.patterns[i]
		matchResult := entry.pattern.Match(components, isDir)

		if matchResult != gitignore.NoMatch {
			lastMatch = entry
			lastMatchResult = matchResult
		}
	}

	// No pattern matched - include by default
	if lastMatch == nil {
		result.Ignored = false
		result.Reason = ReasonNoMatch
		return result
	}

	// Determine result based on match type
	switch lastMatchResult {
	case gitignore.Exclude:
		result.Ignored = true
		if lastMatch.source == SourceDefaults {
			result.Reason = ReasonDefaultDeny
		} else {
			result.Reason = ReasonExcluded
		}
	case gitignore.Include:
		result.Ignored = false
		result.Reason = ReasonIncluded
	default:
		// Shouldn't happen, but handle gracefully
		result.Ignored = false
		result.Reason = ReasonNoMatch
	}

	result.Source = lastMatch.source
	result.Pattern = lastMatch.rawPattern

	return result
}

// Patterns returns all loaded patterns for debugging/dry-run.
// The returned slice is a copy and can be safely modified.
func (m *Matcher) Patterns() []string {
	result := make([]string, len(m.patterns))
	for i, entry := range m.patterns {
		result[i] = fmt.Sprintf("[%s] %s", entry.source, entry.rawPattern)
	}
	return result
}

// PatternCount returns the total number of loaded patterns.
func (m *Matcher) PatternCount() int {
	return len(m.patterns)
}

// RootDir returns the root directory used by this matcher.
func (m *Matcher) RootDir() string {
	return m.rootDir
}

File: backend/libs/ts/plugin-package/src/client/ignore/matcher.ts
// Ignore matcher: composes pattern sources with gitignore precedence (defaults
// → .gitignore → .stigmerignore → caller flags), evaluating "last match wins".
// A faithful port of the Go CLI's pkg/ignore Matcher.
//
// The matcher is pure over the TEXT of the ignore files. Where that text
// comes from is the caller's edge: the CLI reads `.gitignore` and
// `.stigmerignore` from the directory it is about to package, the console
// finds them among the files a marketplace tree lists. Both build the same
// matcher from the same sources, which is what makes a plugin selected in
// the browser the same set of files the CLI would have selected.

import { DEFAULT_PATTERNS } from "./defaults.js";
import { MatchResult, type Pattern, parsePattern } from "./pattern.js";

export const SOURCE_DEFAULTS = "defaults";
export const SOURCE_GITIGNORE = ".gitignore";
export const SOURCE_STIGMERIGNORE = ".stigmerignore";
export const SOURCE_CLI = "cli";

export enum Reason {
  NoMatch = 0,
  Excluded = 1,
  Included = 2,
  DefaultDeny = 3,
}

export const REASON_TEXT: Readonly<Record<Reason, string>> = {
  [Reason.NoMatch]: "no pattern matched",
  [Reason.Excluded]: "excluded by pattern",
  [Reason.Included]: "included by negation pattern",
  [Reason.DefaultDeny]: "excluded by security default",
};

export interface MatchReason {
  readonly ignored: boolean;
  readonly reason: Reason;
  readonly source: string;
  readonly pattern: string;
}

interface PatternEntry {
  readonly pattern: Pattern;
  readonly source: string;
  readonly raw: string;
}

/** The pattern sources a matcher is built from, in precedence order (later wins). */
export interface IgnoreSources {
  /** Whether the built-in security defaults apply (every push says yes). */
  readonly includeDefaults: boolean;
  /** The text of the root `.gitignore`, or `undefined` when absent or not respected. */
  readonly gitignore?: string;
  /** The text of the root `.stigmerignore`, or `undefined` when absent. */
  readonly stigmerignore?: string;
  /** Extra exclusion patterns from the caller (`--ignore`). */
  readonly extraIgnore?: readonly string[];
  /** Extra inclusion patterns from the caller (`--include`), negated if not already. */
  readonly extraInclude?: readonly string[];
}

export class Matcher {
  private readonly entries: readonly PatternEntry[];

  constructor(entries: readonly PatternEntry[]) {
    this.entries = entries;
  }

  /** Decide whether a path (relative, forward-slash) is ignored, with reason. */
  matchWithReason(path: string, isDir: boolean): MatchReason {
    const components = pathToComponents(path);
    if (components.length === 0) {
      return { ignored: false, reason: Reason.NoMatch, source: "", pattern: "" };
    }

    let lastMatch: PatternEntry | undefined;
    let lastResult = MatchResult.NoMatch;
    for (const entry of this.entries) {
      const result = entry.pattern.match(components, isDir);
      if (result !== MatchResult.NoMatch) {
        lastMatch = entry;
        lastResult = result;
      }
    }

    if (lastMatch === undefined) {
      return { ignored: false, reason: Reason.NoMatch, source: "", pattern: "" };
    }

    if (lastResult === MatchResult.Exclude) {
      const reason = lastMatch.source === SOURCE_DEFAULTS ? Reason.DefaultDeny : Reason.Excluded;
      return { ignored: true, reason, source: lastMatch.source, pattern: lastMatch.raw };
    }
    // Include (negation): explicitly un-ignored.
    return { ignored: false, reason: Reason.Included, source: lastMatch.source, pattern: lastMatch.raw };
  }

  /** Whether the path should be excluded from the artifact. */
  match(path: string, isDir: boolean): boolean {
    return this.matchWithReason(path, isDir).ignored;
  }

  /** All loaded patterns as "[source] pattern" (for dry-run diagnostics). */
  patterns(): string[] {
    return this.entries.map((entry) => `[${entry.source}] ${entry.raw}`);
  }
}

/** Build a matcher from its sources; the pure core every reader shares. */
export function buildMatcher(sources: IgnoreSources): Matcher {
  const entries: PatternEntry[] = [];
  if (sources.includeDefaults) {
    entries.push(...parseEntries(DEFAULT_PATTERNS, SOURCE_DEFAULTS));
  }
  if (sources.gitignore !== undefined) {
    entries.push(...parseEntries(sources.gitignore.split(/\r?\n/), SOURCE_GITIGNORE));
  }
  if (sources.stigmerignore !== undefined) {
    entries.push(...parseEntries(sources.stigmerignore.split(/\r?\n/), SOURCE_STIGMERIGNORE));
  }

  const extraIgnore = sources.extraIgnore ?? [];
  const extraInclude = sources.extraInclude ?? [];
  if (extraIgnore.length > 0 || extraInclude.length > 0) {
    entries.push(...loadCliPatterns(extraIgnore, extraInclude));
  }

  return new Matcher(entries);
}

function parseEntries(patterns: readonly string[], source: string): PatternEntry[] {
  const entries: PatternEntry[] = [];
  for (const raw of patterns) {
    const trimmed = raw.trim();
    if (trimmed === "" || trimmed.startsWith("#")) continue;
    entries.push({ pattern: parsePattern(trimmed, []), source, raw: trimmed });
  }
  return entries;
}

function loadCliPatterns(ignore: readonly string[], include: readonly string[]): PatternEntry[] {
  const entries = parseEntries(ignore, SOURCE_CLI);
  const negations: string[] = [];
  for (const pattern of include) {
    const trimmed = pattern.trim();
    if (trimmed === "") continue;
    negations.push(trimmed.startsWith("!") ? trimmed : `!${trimmed}`);
  }
  entries.push(...parseEntries(negations, SOURCE_CLI));
  return entries;
}

function pathToComponents(path: string): string[] {
  const normalized = path.replace(/\\/g, "/").replace(/^\.\//, "");
  if (normalized === "" || normalized === ".") return [];
  return normalized.split("/");
}
